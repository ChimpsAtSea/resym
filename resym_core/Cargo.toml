[package]
name = "resym_core"
version.workspace = true
edition = "2021"
rust-version = "1.75"
exclude = [
    "tests/data/*",
    "tests/snapshots/*",
]

[features]
default = ["rayon", "http"]

rayon = ["dep:rayon", "dep:num_cpus", "dashmap/rayon"]
http = ["dep:ehttp", "dep:url"]

[dependencies]
pdb = { git = "https://github.com/ergrelet/pdb", branch = "0.8.0-resym" }
thiserror = "1.0"
log = "0.4"
rayon = { version = "1.5", optional = true }
num_cpus = { version = "1.13", optional = true }
crossbeam-channel = "0.5"
dashmap = { version = "5.2" }
regex = "1.5"
similar = "2.1"
ehttp = { version = "0.2", optional = true }
url = { version = "2.3", optional = true }

# Web:
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm_thread = { version = "0.2", features = ["es_modules"] }
instant = { version = "0.1", features = ["wasm-bindgen"] }

[dev-dependencies]
insta = "1.14"
