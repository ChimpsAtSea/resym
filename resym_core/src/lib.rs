pub mod backend;
pub mod diffing;
pub mod frontend;
pub mod pdb_file;
pub mod pdb_types;
pub mod syntax_highlighting;

const PKG_VERSION: &str = env!("CARGO_PKG_VERSION");
