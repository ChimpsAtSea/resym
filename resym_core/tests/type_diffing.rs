use std::path::Path;

use resym_core::{
    diffing::diff_type_by_name, pdb_file::PdbFile, pdb_types::PrimitiveReconstructionFlavor,
};

const TEST_PDB_FROM_FILE_PATH: &str = "tests/data/test_diff_from.pdb";
const TEST_PDB_TO_FILE_PATH: &str = "tests/data/test_diff_to.pdb";
const TEST_CASES: &[&str] = &[
    "UserStructAddAndReplace",
    "UserStructRemove",
    "UserStructAdd",
    "RemovedStruct",
    "NewStruct",
];

#[test]
fn test_struct_diffing() {
    let pdb_file_from = PdbFile::load_from_file(Path::new(TEST_PDB_FROM_FILE_PATH))
        .expect("load test_diff_from.pdb");
    let pdb_file_to =
        PdbFile::load_from_file(Path::new(TEST_PDB_TO_FILE_PATH)).expect("load test_diff_to.pdb");

    for test_case_type_name in TEST_CASES {
        let diffed_type = diff_type_by_name(
            &pdb_file_from,
            &pdb_file_to,
            test_case_type_name,
            PrimitiveReconstructionFlavor::Portable,
            false,
            false,
            false,
            false,
        )
        .expect("diff generation");
        insta::assert_snapshot!(diffed_type.data);
    }
}

#[test]
fn test_struct_diffing_inexistent_type() {
    const INEXISTENT_TYPE_NAME: &str = "TypeNotFound";
    let pdb_file_from = PdbFile::load_from_file(Path::new(TEST_PDB_FROM_FILE_PATH))
        .expect("load test_diff_from.pdb");
    let pdb_file_to =
        PdbFile::load_from_file(Path::new(TEST_PDB_TO_FILE_PATH)).expect("load test_diff_to.pdb");
    assert!(diff_type_by_name(
        &pdb_file_from,
        &pdb_file_to,
        INEXISTENT_TYPE_NAME,
        PrimitiveReconstructionFlavor::Portable,
        false,
        false,
        false,
        false,
    )
    .is_err());
}
