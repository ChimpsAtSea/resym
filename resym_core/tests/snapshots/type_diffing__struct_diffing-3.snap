---
source: resym_core/tests/type_diffing.rs
expression: diffed_type.data
---
 
-struct UserStructAdd { /* Size=0x10 */
+struct UserStructAdd { /* Size=0x18 */
   /* 0x0000 */ int32_t field1;
   /* 0x0008 */ void* field2;
+  /* 0x0010 */ char field3;
+  /* 0x0014 */ int32_t field4;
 };

